File: src/ParentsImputerByProgeny.cpp
#include <cmath>
#include "../include/ParentsImputerByProgeny.h"
#include "../include/common.h"

using namespace std;

ParentsImputerByProgeny::ParentsImputerByProgeny(
							const vector<VCFFamilyRecord *>& rs,
							const vector<vector<int>>& ref_hs_mat,
							const vector<vector<int>>& ref_hs_pat,
							const Map& map_, double w) :
					VCFHMM(map_, w), records(rs),
					ref_haps_mat(ref_hs_mat),
					ref_haps_pat(ref_hs_pat),
					prev_h_table(collect_possible_previous_hidden_states()),
					Cc(calc_Cc(rs)), Cp(calc_Cp(rs)) { }

vector<double> ParentsImputerByProgeny::calc_Cc(
							const vector<VCFFamilyRecord *>& rs) const {
	const size_t	M = rs.size();
	vector<double>	Cc(M-1);
	for(size_t i = 0; i < M - 1; ++i) {
		Cc[i] = Map::Kosambi(dist(rs[i], rs[i+1]));
	}
	return Cc;
}

vector<double> ParentsImputerByProgeny::calc_Cp(
							const vector<VCFFamilyRecord *>& rs) const {
	const double	K = 5.0;
	const size_t	M = rs.size();
	vector<double>	Cp(M-1);
	for(size_t i = 0; i < M - 1; ++i) {
		Cp[i] = Map::Kosambi(dist(rs[i], rs[i+1]) * K);
	}
	return Cp;
}

pair<int, int> ParentsImputerByProgeny::compute_parent_gt(int i, int h) const {
	auto parent_gt = [&](int hc, int hp, int prog_a, bool maternal) -> int {
		const auto& ref_haps = maternal ? ref_haps_mat : ref_haps_pat;
		
		if(hc == 0)
			return prog_a | (ref_haps[hp][i] << 1);
		else
			return ref_haps[hp][i] | (prog_a << 1);
	};
	
	const int	hw = h & 1;				// Progeny haplotype inherited from mat
	const int	hc1 = (h >> 1) & 1;		// Maternal haplotype passed to progeny
	const int	hc2 = (h >> 2) & 1;		// Paternal haplotype passed to progeny
	const int	hp1 = (h >> 3) % NH();	// Maternal haplotype not passed on
	const int	hp2 = (h >> 3) / NH();	// Paternal haplotype not passed on
	
	// Allele transmitted from the parents
	const int prog_allele1 = records[i]->get_allele(2, hw);
	const int prog_allele2 = records[i]->get_allele(2, 1 - hw);
	
	return make_pair(
		parent_gt(hc1, hp1, prog_allele1, true),
		parent_gt(hc2, hp2, prog_allele2, false)
	);
}

double ParentsImputerByProgeny::emission_probability(std::size_t i,
															int h) const {
	const auto	p = compute_parent_gt(i, h);
	const int	mat_phased_gt = p.first;
	const int	pat_phased_gt = p.second;
	const int	mat_gt = records[i]->unphased_mat();
	const int	pat_gt = records[i]->unphased_pat();
	return E[mat_phased_gt][mat_gt] + E[pat_phased_gt][pat_gt];
}

double ParentsImputerByProgeny::transition_probability(size_t i,
													int prev_h, int h) const {
	const double	cc = Cc[i-1];	// progeny transition probability
	const double	cp = Cp[i-1];	// parental transition probability
	const int	prev_hc1 = (prev_h >> 1) & 1;
	const int	prev_hc2 = (prev_h >> 2) & 1;
	const int	prev_hp1 = (prev_h >> 3) % NH();
	const int	prev_hp2 = (prev_h >> 3) / NH();
	const int	hc1 = (h >> 1) & 1;
	const int	hc2 = (h >> 2) & 1;
	const int	hp1 = (h >> 3) % NH();
	const int	hp2 = (h >> 3) / NH();
	
	return log(prev_hc1 != hc1 ? cc : 1.0 - cc) +
		   log(prev_hc2 != hc2 ? cc : 1.0 - cc) +
		   log(prev_hp1 != hp1 ? cp : 1.0 - cp) +
		   log(prev_hp2 != hp2 ? cp : 1.0 - cp);
}

vector<ParentsImputerByProgeny::DP>
			ParentsImputerByProgeny::initialize_dp() const {
	const size_t	L = num_states();
	vector<DP>	dp(M(), DP(L, pair<double, int>(MIN_PROB, 0)));
	for(int h = 0; h < (int)L; ++h) {	// hidden state
		const double	E_all = emission_probability(0, h);
		dp[0][h] = make_pair(E_all, h);
	}
	return dp;
}

// Collect possible previous hidden states for the hidden state.
vector<vector<int>>
ParentsImputerByProgeny::collect_possible_previous_hidden_states() const {
	const int L = num_states();
	vector<vector<int>> prev_h_table(L);
	
	for(int h = 0; h < L; ++h) {
		const int	hw  = h & 1;
		const int	hc1 = (h >> 1) & 1;
		const int	hc2 = (h >> 2) & 1;
		const int	hp1 = (h >> 3) % NH();
		const int	hp2 = (h >> 3) / NH();
		
		// Assume that at most one component can switch simultaneously
		for(int prev_h = hw; prev_h < L; prev_h += 2) {
			const int	prev_hc1 = (prev_h >> 1) & 1;
			const int	prev_hc2 = (prev_h >> 2) & 1;
			const int	prev_hp1 = (prev_h >> 3) % NH();
			const int	prev_hp2 = (prev_h >> 3) / NH();
			
			int counter = static_cast<int>(prev_hc1 != hc1) +
						  static_cast<int>(prev_hc2 != hc2) +
						  static_cast<int>(prev_hp1 != hp1) +
						  static_cast<int>(prev_hp2 != hp2);
			if(counter <= 1)
				prev_h_table[h].push_back(prev_h);
		}
	}
	
	return prev_h_table;
}

void ParentsImputerByProgeny::update_dp(size_t i, vector<DP>& dp) const {
	const int	L = num_states();
	for(int h = 0; h < L; ++h) {	// hidden state
		// Emission log-probability for state h at position i
		const double	E_all = emission_probability(i, h);
		for(int prev_h : prev_h_table[h]) {
			// Transition log-probability from prev_h -> h
			const double	Tp = transition_probability(i, prev_h, h);
			const double	prob = dp[i-1][prev_h].first + (Tp + E_all);
			dp[i][h] = max(dp[i][h], make_pair(prob, prev_h));
		}
	}
}

void ParentsImputerByProgeny::update_genotypes(const vector<int>& hs) {
	for(size_t i = 0; i < this->M(); ++i) {
		VCFFamilyRecord	*record = records[i];
		const auto	p = compute_parent_gt(i, hs[i]);
		const int	mat_gt = p.first;
		const int	pat_gt = p.second;
		record->set_geno(0, mat_gt | 4);
		record->set_geno(1, pat_gt | 4);
	}
}

void ParentsImputerByProgeny::impute() {
	// DP
	vector<DP>	dp = initialize_dp();
	for(int i = 1; i < (int)this->M(); ++i) {
		update_dp(i, dp);
	}
	
	const vector<int>	hs = trace_back(dp);
	update_genotypes(hs);
}

File: python/ParentsImputerByProgeny.py
from __future__ import annotations

# coding: utf-8
# ParentsImputerByProgeny.py
# 後代がimputedで片親がknownな家系の補完
# 後代が複数だと難しいので一つでimputedとする

from math import log
from typing import List, Tuple

from VCFFamily import VCFFamilyRecord
from VCFHMM import *
from Genotype import Genotype


#################### ParentsImputerByProgeny ####################

MIN_PROB = -1e300

class ParentsImputerByProgeny(VCFHMM[VCFFamilyRecord]):
	DP = List[Tuple[float, int]]	# (log of probability, prev h)
	
	def __init__(self, records: list[VCFFamilyRecord],
								ref_haps_mat: list[list[int]],
								ref_haps_pat: list[list[int]],
								map_: Map, w: float):
		VCFHMM.__init__(self, records, map_)
		self.records: list[VCFFamilyRecord] = records
		self.ref_haps_mat = ref_haps_mat
		self.ref_haps_pat = ref_haps_pat
		self.prev_h_table = self.collect_possible_previous_hidden_states()
		# crossover values
		# 後代
		self.Cc = [ Map.Kosambi(self.dist(r1, r2))
							for r1, r2 in zip(records, records[1:]) ]
		K = 5						# 親の遷移確率の子に対する倍率
		# 親は直接の後代でないので、乗り換え確率を高くする
		self.Cp = [ Map.Kosambi(self.dist(r1, r2) * K)
							for r1, r2 in zip(self.records, self.records[1:]) ]
	
	def NH(self) -> int:
		return len(self.ref_haps_mat)
	
	def num_states(self) -> int:
		return self.NH()**2 << 3
	
	# ハプロタイプを決めたときのGenotype
	def gt_by_haplotypes(self, hc1: int, hc2: int,
								mat_gt: int, pat_gt: int) -> int:
		return ((mat_gt >> hc1) & 1) | (((pat_gt >> hc2) & 1) << 1)
	
	def decode_state(self, h: int) -> tuple[int, int, int, int, int]:
		hw = h & 1			# 後代のどちらに母親から渡ったか
		hc1 = (h >> 1) & 1	# 母親のどちらから後代に渡ったか
		hc2 = (h >> 2) & 1	# 父親のどちらから後代に渡ったか
		hp2, hp1 = divmod(h >> 3, self.NH())	# 親の後代に渡ってこない方
		return (hw, hc1, hc2, hp1, hp2)
	
	def compute_parent_gt(self, i: int, h: int) -> tuple[int, int]:
		def parent_gt(hc: int, hp: int, prog_a: int, b: bool) -> int:
			ref_haps = self.ref_haps_mat if b else self.ref_haps_pat
			if hc == 0:
				return prog_a | (ref_haps[hp][i] << 1)
			else:
				return ref_haps[hp][i] | (prog_a << 1)
		
		hw, hc1, hc2, hp1, hp2 = self.decode_state(h)
		prog_allele1 = self.records[i].get_allele(2, hw)	# 親から渡ったアレル
		prog_allele2 = self.records[i].get_allele(2, 1-hw)	# 親から渡ったアレル
		return (parent_gt(hc1, hp1, prog_allele1, True),
				parent_gt(hc2, hp2, prog_allele2, False))
	
	# i: record index
	def emission_probability(self, i: int, h: int) -> float:
		mat_phased_gt, pat_phased_gt = self.compute_parent_gt(i, h)
		mat_gt = self.records[i].unphased_mat()
		pat_gt = self.records[i].unphased_pat()
		return self.E[mat_phased_gt][mat_gt] + self.E[pat_phased_gt][pat_gt]
	
	def transition_probability(self, i: int, prev_h: int, h: int) -> float:
		cc = self.Cc[i-1]	# 後代の遷移確率
		cp = self.Cp[i-1]	# 親の遷移確率
		hw, hc1, hc2, hp1, hp2 = self.decode_state(h)
		_, prev_hc1, prev_hc2, prev_hp1, prev_hp2 = self.decode_state(prev_h)
		
		# 遷移確率 0なら(1-c)、1ならcを掛ける
		return (log(cc if prev_hc1 != hc1 else 1.0 - cc) +
				log(cc if prev_hc2 != hc2 else 1.0 - cc) +
				log(cp if prev_hp1 != hp1 else 1.0 - cp) +
				log(cp if prev_hp2 != hp2 else 1.0 - cp))
	
	def initialize_dp(self) -> list[DP]:
		M = len(self.records)	# マーカー数
		L = self.num_states()
		dp = [ [ (MIN_PROB, 0) ] * L for _ in range(M) ]
		for h in range(L):		# hidden state
			E_all = self.emission_probability(0, h)
			dp[0][h] = (E_all, h)
		return dp
	
	# hidden stateに対して、可能な前のhidden stateを集めておく
	def collect_possible_previous_hidden_states(self) -> list[list[int]]:
		L = self.num_states()
		prev_h_table: list[list[int]] = [ [] for _ in range(L) ]
		for h in range(L):		# hidden state
			hw, hc1, hc2, hp1, hp2 = self.decode_state(h)
			# 複数が同時に乗り換えることはないとする
			for prev_h in range(hw, L, 2):
				_, prev_hc1, prev_hc2, prev_hp1, prev_hp2 = \
											self.decode_state(prev_h)
				counter = (int(prev_hc1 != hc1) +
						   int(prev_hc2 != hc2) +
						   int(prev_hp1 != hp1) +
						   int(prev_hp2 != hp2))
				if counter <= 1:
					prev_h_table[h].append(prev_h)
		
		return prev_h_table
	
	def update_dp(self, i: int, dp: list[DP]) -> None:
		L = self.num_states()
		for h in range(L):		# hidden state
			E_all = self.emission_probability(i, h)
			for prev_h in self.prev_h_table[h]:
				Tp = self.transition_probability(i, prev_h, h)
				prob = dp[i-1][prev_h][0] + (Tp + E_all)
				dp[i][h] = max(dp[i][h], (prob, prev_h))
	
	def update_genotypes(self, hs: list[int]) -> None:
		M = len(self.records)
		for i in range(M):
			record = self.records[i]
			mat_gt, pat_gt = self.compute_parent_gt(i, hs[i])
			record.geno[0] = mat_gt | 4
			record.geno[1] = pat_gt | 4
	
	def impute(self) -> None:
		# DP
		M = len(self.records)
		dp = self.initialize_dp()
		for i in range(1, M):
			self.update_dp(i, dp)
		
		hs = self.trace_back(dp)
		self.update_genotypes(hs)
